freeipa/src/main/java/com/sequenceiq/freeipa/sync/FreeipaChecker.java

```diff
@@ -31,28 +34,38 @@
     @Inject
     private FreeIpaClientFactory freeIpaClientFactory;
 
-    private RPCResponse<Boolean> checkStatus(Stack stack, Set<InstanceMetaData> checkableInstances) throws Exception {
+    @Inject
+    private InstanceMetaDataService instanceMetaDataService;
+
+    private List<RPCResponse<Boolean>> checkStatus(Stack stack, Set<InstanceMetaData> checkableInstances) throws Exception {
         return checkedMeasure(() -> {
-            FreeIpaClient freeIpaClient = checkedMeasure(() -> freeIpaClientFactory.getFreeIpaClientForStackWithPing(stack), LOGGER,
-                    ":::Auto sync::: freeipa client is created in {}ms");
-            String hostname = getPrimaryHostname(checkableInstances);
-            return checkedMeasure(() -> freeIpaClient.serverConnCheck(freeIpaClient.getHostname(), hostname), LOGGER,
-                    ":::Auto sync::: freeipa server_conncheck ran in {}ms");
+            List<RPCResponse<Boolean>> statuses = new LinkedList<>();
+            for (InstanceMetaData instanceMetaData : checkableInstances) {
+                String hostname = instanceMetaData.getDiscoveryFQDN();
+                FreeIpaClient freeIpaClient = checkedMeasure(() ->
+                                freeIpaClientFactory.getFreeIpaClientForStackWithPing(stack, hostname), LOGGER,
+                        ":::Auto sync::: freeipa client is created in {}ms");
+                statuses.add(checkedMeasure(() -> freeIpaClient.serverConnCheck(freeIpaClient.getHostname(), hostname), LOGGER,
+                        ":::Auto sync::: freeipa server_conncheck ran in {}ms"));
+            }
+            return statuses;
         }, LOGGER, ":::Auto sync::: freeipa server status is checked in {}ms");
     }
 
     public SyncResult getStatus(Stack stack, Set<InstanceMetaData> checkableInstances) {
         try {
-            RPCResponse<Boolean> response = checkStatus(stack, checkableInstances);
+            Set<InstanceMetaData> notTermiatedStackInstances = instanceMetaDataService.findNotTerminatedForStack(stack.getId());
+            List<RPCResponse<Boolean>> responses = checkStatus(stack, checkableInstances);
             DetailedStackStatus status;
             String postFix = "";
-            if (response.getResult()) {
+            Boolean result = !responses.isEmpty() && responses.stream().allMatch(RPCResponse::getResult);
+            if (result && responses.size() == notTermiatedStackInstances.size()) {
                 status = DetailedStackStatus.PROVISIONED;
             } else {
                 status = DetailedStackStatus.UNHEALTHY;
                 postFix = "Freeipa is unhealthy, ";
             }
-            return new SyncResult(postFix + getMessage(response), status, response.getResult());
+            return new SyncResult(postFix + getMessages(responses), status, result);
         } catch (HttpHostConnectException | ConnectTimeoutException e) {
```

Comment:
is this still a valid option? can we receive such exceptions?

Reply:
I noticed this yesterday too. You are correct, it is impossible to receive these exceptions. I am in the process of fixing this in CB-6745. I would prefer to fix this in a separate pull request since this wasn't affected by this pull request and this PR is pretty large already.